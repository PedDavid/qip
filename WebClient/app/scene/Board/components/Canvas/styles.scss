canvas.debug {
    height: 800px;
    width: 1200px;
    border: 1px solid #000000;
    position: fixed;
    left: 0;
    top: 0;
}