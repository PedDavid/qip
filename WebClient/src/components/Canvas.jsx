// @flow

import React from 'react'

type PointerEvents = {
  onDown?: any => void,
  onUp?: any => void,
  onMove?: any => void,
  onOut?: any => void
}

export default class Canvas extends React.Component {
  props: PointerEvents
  canvas: window.EventListener

  componentDidMount () {
    if (this.props.onDow) {
      this.canvas.addEventListener('pointerdown', this.props.onDown, false)
    }
    if (this.props.onUp) {
      this.canvas.addEventListener('pointerup', this.props.onUp, false)
    }
    if (this.props.onMove) {
      this.canvas.addEventListener('pointermove', this.props.onMove, false)
    }
    if (this.props.onOut) {
      this.canvas.addEventListener('pointerout', this.props.onOut, false)
    }
  }

  render () {
    const { onDown, onUp, onMove, onOut, ...props } = this.props
    const style = { border: '1px solid #000000' } // TODO(PedDavid): Delete this, used only for debug purposes
    return <canvas ref={canvas => { this.canvas = canvas }} {...props} style={style} />
  }
}
