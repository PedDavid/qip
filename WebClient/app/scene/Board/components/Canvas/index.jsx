import React from 'react'
import styles from './styles.scss'

export default class Canvas extends React.Component {
  componentDidMount () {
    if (this.props.onDown != null) {
      this.canvas.addEventListener('pointerdown', this.props.onDown, false)
    }
    if (this.props.onUp != null) {
      this.canvas.addEventListener('pointerup', this.props.onUp, false)
    }
    if (this.props.onMove != null) {
      this.canvas.addEventListener('pointermove', this.props.onMove, false)
    }
    if (this.props.onOut != null) {
      this.canvas.addEventListener('pointerout', this.props.onOut, false)
    }
  }

  shouldComponentUpdate () {
    return false
  }

  render () {
    const {onDown, onUp, onMove, onOut, ...props} = this.props
    return <canvas ref={(canvas) => { this.canvas = canvas }} className={styles.debug} {...props} />
  }
}
