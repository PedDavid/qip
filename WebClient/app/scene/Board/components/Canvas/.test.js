import React from 'react'
import { shallow, mount } from 'enzyme'

import Canvas from '.'

describe('Component: Canvas', () => {
  const props = { width: 1, height: 1 }
  const wrapper = shallow(<Canvas {...props} />)
  it('Is not null', () => {
    expect(wrapper).not.toBeNull()
  })
  it('Is instance of Canvas', () => {
    expect(wrapper.instance()).toBeInstanceOf(Canvas)
  })
  it('Type is <canvas>', () => {
    expect(wrapper.type()).toBe('canvas')
  })
  it('Props are delivered', () => {
    expect(wrapper.instance().props).toMatchObject(props)
  })
  it('Renders as expected', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  // Otherwise browsers throw warnings because prop is unknown to <canvas>
  it('Does not pass listeners to <canvas>', () => {
    const listeners = { onDown: evt => {} }
    const wrapper = shallow(<Canvas {...props} {...listeners} />)
    expect(wrapper.instance().props).toHaveProperty('onDown')
    expect(wrapper.props()).not.toHaveProperty('onDown')
  })
  it('Receives all listeners', () => {
    const listeners = {
      onDown: evt => {},
      onUp: evt => {},
      onMove: evt => {},
      onOut: evt => {}
    }
    const wrapper = shallow(<Canvas {...props} {...listeners} />)
    expect(wrapper).not.toBeNull()
  })
  it('Mounts', () => {
    const wrapper = mount(<Canvas {...props} />)
    expect(wrapper).not.toBeNull()
  })
  it('Updates and does not change state', () => {
    const wrapper = mount(<Canvas {...props} />)
    const updated = wrapper.update()
    expect(updated).toMatchObject(wrapper)
  })
  it(`Should test listeners, sadly enzyme can't simulate pointer Events :(
    Right now it only insures test coverage is 100%`, () => {
    const listeners = {
      onDown: evt => {},
      onUp: evt => {},
      onMove: evt => {},
      onOut: evt => {}
    }
    const wrapper = mount(<Canvas {...props} {...listeners} />)
    expect(wrapper).not.toBeNull()
    //TODO(peddavid): Simulate pointer events
  })
})
