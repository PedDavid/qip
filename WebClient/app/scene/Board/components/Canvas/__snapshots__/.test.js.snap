// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Canvas Renders as expected 1`] = `"<canvas class=\\"debug\\" width=\\"1\\" height=\\"1\\"></canvas>"`;
